### PR TITLE
feat: Using NPM Workspaces and Docker updates to enhance dev usability

### PR DESCRIPTION
**Pull request by elliot-sabitov**
feat: Utilizing npm workspaces to simplify multi npm package local dependencies
feat: Adjusting Docker related files to expose debugger port
chore: Created run configuration support files to run Docker and attach debugger
docs: Updated REAME with relevant information

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,29 +41,26 @@ Before you begin, make sure you have the following installed on your system:
...
-1. Install project dependencies:
+1. Install project dependencies from root dir:
 
    ```shell
-   ./unix-init-install-all.sh
+   npm run install-all
    ```
 
-1. Start the server and its supporting infrastructure with:
+1. The docker container should be initialized by running `docker-compose -f ./Server/docker-compose.yml up -d` or
```
Comment (lydiazcheng) [inline README.md:50]: This command does not work. The yml file seems to be moved into root folder.

File: README.md [hunk 2]
```diff
@@ -41,29 +41,26 @@ Before you begin, make sure you have the following installed on your system:
...
+1. Install project dependencies from root dir:
 
    ```shell
-   ./unix-init-install-all.sh
+   npm run install-all
    ```
 
-1. Start the server and its supporting infrastructure with:
+1. The docker container should be initialized by running `docker-compose -f ./Server/docker-compose.yml up -d` or 
+by using the IntelliJ `Server` Run Configuration which was created for this purpose.
```
Comment (lydiazcheng) [inline README.md:51]: Sorry due to my very limited knowledge on typescript project, I am not exactly sure about what need to be done when reading this sentence. Would you mind explaining more for this part?

File: README.md [hunk 3]
```diff
@@ -41,29 +41,26 @@ Before you begin, make sure you have the following installed on your system:
...
 
-    ```shell
-    docker-compose up -d 
-    ```
+1. Running `docker-compose.yml` will ensure that the container is configured to expose the `:9229` debugging
+port for the underlying NodeJS process. A variety of tools can be utilized to establish a debugger connection
+with the exposed localhost 9229 port which is forwarded to the NodeJS service running within docker. The IntelliJ
+`Attach Debugger` Run Configuration was made to attach to a debugging session.
```
Comment (lydiazcheng) [inline README.md:56]: Does this mean create a new run configuration and select "Attach to Node.js/Chrome"?

**Review by lydiazcheng — COMMENTED** [review]
Hi Elliot, the change looks good to me but I have some questions. Could you please help?

File: Server/src/index.ts [hunk 4]
```diff
@@ -3,169 +3,319 @@
...
+    addFormats(this._ajv, {
+      mode: 'fast',
+      formats: ['date-time']
+    })
+  }
+
+  private initLogger() {
+    this._logger = new Logger<ILogObj>({
+      name: 'CitrineOS Logger',
+      minLevel: systemConfig.server.logLevel,
```
Comment (venus-yu) [inline Server/src/index.ts:138]: i'm receiving this error when trying to get the docker running: 

`Server/src/index.ts(138,30): error TS2339:  Property 'server' does not exist on type '{ env: "development" | "production"; data:
 { sequelize: { host: string; port: number; database: string; username: string; password: string; storage: string; sync: boolean; dia
lect?: any; }; }; centralSystem: { ...; }; ... 4 more ...; maxCachingSeconds: number; }'`

i locally updated this to be `systemConfig.logLevel` and the docker containers ran successfully - should this same change be made to the PR as well or is this unique to my machine?

File: Server/src/index.ts [hunk 5]
```diff
@@ -3,169 +3,319 @@
...
+  }
 
-        // Shut down all modules and central system
-        this._modules.forEach(module => {
-            module.shutdown();
-        });
-        this._centralSystem.shutdown();
+  private initCentralSystem() {
+    this._centralSystem = new CentralSystemImpl(this._config, this._cache as ICache, undefined, undefined, this._logger, this._ajv)
+    this.host = this._config.server.host
```
Comment (venus-yu) [inline Server/src/index.ts:174]: i'm receiving this additional error while trying to get the docker running:

`Server/src/index.ts(174,30): error TS2339: Property 'server' does not exist on type '{ env: "development" | "production"; data: { sequelize: { host: string; port: number; database: string; username: string; password: string; storage: string; sync: boolean; dialect?: any; }; }; centralSystem: { ...; }; ... 4 more ...; maxCachingSeconds: number; }'`

i locally updated this to be `this.host = this._config.centralSystem.host` and `this.port = this._config.centralSystem.port` and the docker containers ran successfully - should this same change be made to the PR as well or is this unique to my machine?

File: README.md (hunk 1 above)
Comment (elliot-sabitov) [reply README.md]: yep, I just updated the readme

File: README.md (hunk 2 above)
Comment (elliot-sabitov) [reply README.md:51]: @lydiazcheng take a look here https://www.jetbrains.com/help/idea/run-debug-configuration.html#templates it is particular to Intellij but it creates the runnable jobs in the IDE and the jobs essentially run various terminal commands to run docker or npm or attach debugger, etc.

File: README.md (hunk 3 above)
Comment (elliot-sabitov) [reply README.md:56]: @lydiazcheng yes I forgot to commit the run configurations, but they should be there now. CC: @venus-yu

File: Server/src/index.ts (hunk 4 above)
Comment (elliot-sabitov) [reply Server/src/index.ts]: @venus-yu great catch, this was something that I pulled from latest changes from Thana that was incorrectly merged, I resolved this

File: Server/src/index.ts (hunk 5 above)
Comment (elliot-sabitov) [reply Server/src/index.ts]: @venus-yu should be fixed

**Review by thanaParis — CHANGES_REQUESTED** [review]
Please remove the .idea folder and add it to the gitignore. we should not add something specific to IntelliJ. I think it would make sense to either remove mention of IntelliJ from the README, or add a description of the setup in VSCode and an open invitation to add instructions for other IDEs--let me know what you think, I could possibly add the VSCode advice if we feel it's important to keep instructions on how to attach a debugger (that may be too extraneous to the project to keep).

**Comment by elliot-sabitov** [comment]
> Please remove the .idea folder and add it to the gitignore. we should not add something specific to IntelliJ. I think it would make sense to either remove mention of IntelliJ from the README, or add a description of the setup in VSCode and an open invitation to add instructions for other IDEs--let me know what you think, I could possibly add the VSCode advice if we feel it's important to keep instructions on how to attach a debugger (that may be too extraneous to the project to keep).

@thanaParis in this case the .idea folder contains run configurations that make it easier to run the app in IntelliJ. I would say that it would make sense to do the same for vscode. Ideally, devs can come in and hit the ground running

**Comment by thanaParis** [comment]
Closing this PR for now. 
Moving to an npm workspace is definitely a priority, however, there are some issues outstanding here:

1. There should not be docker configuration or server code at root. That belongs in the appropriate implementation package, such as Server.
2. We won't be merging IntelliJ-specific configuration to the code. I understand that it could be useful to know about for those who use that IDE; feel free to add IntelliJ advice to the README or github.io.

**Pull request closed without merging**